sec2_the_ruby_programming_language/2.2_string.rb

```diff
@@ -0,0 +1,73 @@
+#Strings Part 1 and Part 2
+
+#String concatenation
+
+first_name = "Zi Qiang"
+last_name = "Lim"
+full_name = first_name + " " + last_name 
+puts full_name
+
+#String interpolation
+#Note: Single quote cannot be used in string interpolation
+puts "My first name is #{first_name} and my last_name is #{last_name}" #will print Zi Qiang as first name and Lim as last name
+puts 'My first name is #{first_name} and my last_name is #{last_name}' # will print #{first_name} as first name and #{last_name} as last name instead
+
+#Using methods
+puts full_name.class #String
+#puts fullname.methods #check mehtods available for class type by default
+puts 10.class        #Integer
+#Method chaining(converting a 10 to a string)
+puts 10.to_s.class   #String  
+
+puts full_name.length       #12
+puts full_name.reverse      #miL gnaiQ iZ
+puts full_name.capitalize   #Zi qiang lim
+puts full_name.empty?       #false
+puts "".empty?              #true
+puts "".nil?                #false
+puts nil.nil?               #true
+
+#sub vs gsub
+puts sentence = "Welcome to the jungle"     #Welcome to the jungle
+#sub replace single word
+puts sentence.sub("the jungle","utopia")    #Welcome to utopia
+#Note: gsub replace all of the word in the paragraph
+puts sentence.gsub("the jungle","utopia")   #Welcome to utopia
+
+#Variable assignment
+new_first_name = first_name 
+puts new_first_name #Zi Qiang
+
+first_name = "John" #reassigning first_name to John
+puts first_name #John
+
+#Escaping
+puts 'the new first name is #{new_first_name}'          #the new first name is #{new_first_name}
+puts "the new ifrst name is \#{new_first_name}"         #the new ifrst name is #{new_first_name}
+puts 'Zi Qiang asked \'Hey John, how are you doing?\''  #Zi Qiang asked 'Hey John, how are you doing?'
+
+#Input
+puts "What is your first name?"
+puts "Your first name is #{first_name}"
+puts "Enter a number to multiply by 2"
+
+input = gets.chomp #input is a string in this case
+puts input * 2 #55
+puts input.to_i * 2 #10
```

Comment:
explain this. why?

Reply:
The variable named 'input' with datatype string is multiplied by 2 so another 5 will be added.

When the input.to_i is used, it will convert the input into integer and a result of 10 will be obtained when multiplied by 2.

Reply:
1. Why string multiply by 2, 5 will be added??
2. This seems like normal maths (with to_i). with with string it produces different result?

Reply:
1. The string will be duplicated and appended again at the end.
2. When to_i is used then it will become a normal integer and hence it will multiply the numbers instead of duplicating it.

Reply:
Ok.
So, the question is, how does multiplying string with integer works in ruby?

Reply:
When multiplying a string with an integer, the string will be concatenated with itself.

Reply:
Why? & how?

Reply:
I have tried searching for several websites and there is one showing the function that receives both string and integer as a parameter. The function will add the string(s) according to the number that has passed to it. Eg: when the string "Hello" is multiplied by integer 5, then both "Hello" and 5 will be passed to the function and the Hello will be duplicated 5 times.

Reply:
Why are you using functions to explain a mathematical equation here in ruby? a function is a function. It could have its own implementation. 

Here there is no function being used. Its just plain `"hello" * 5`.

Reply:
I guess it has something to do with the class and object.  Like the string object is being called twice when multiplied by 2. Couldn't really find any clues from the website.

Reply:
@ziqiang2344 😢 I ask you cause I want you to go find out. Not guess guess.

Reply:
Thanks for the advice. I will try my best again to find the answer.

Reply:
When multiplying a string with a number, for example, in this scenario:
`"ABC" * 3 `, it will concatenate 3 instances of itself by carrying out this operation. `"ABC" + "ABC" + "ABC"`.

Besides, if I use + operator on string:
`"ABC" +"ABC"` , the result will be `"ABCABC"`

However, when I use a different operator on string, like below:
`puts "ABC" /2`
`puts "ABC" -"ABC"`

It will show NoMethodError.

This is what I can discover so far. But I have been thinking for a few days still unable to understand why it can be like that.  Would appreciate it if I could get the hints.

Reply:
The answer can be seen in the ruby docs. 
These kind of questions are best answered by referring to the language documentation.
https://ruby-doc.org/core-3.1.1/String.html#2A-method

Reply:
Ah I see, I found this website that time but I explained it as a function.  :( 
Thanks for showing it to me anyway.